packages/tokenbank_v1/contracts/modules/loopring/ExchangeModule.sol

```diff
@@ -0,0 +1,247 @@
+
+    function createOrUpdateDEXAccount(
+        address payable wallet,
+        IExchangeV3     exchange,
+        uint            pubKeyX,
+        uint            pubKeyY,
+        bytes calldata  permission,
+        address         feeToken,
+        uint            feeAmount
+        )
+        external
+        nonReentrantExceptFromThis
+        onlyFromMetaTxOrWalletOwner(wallet)
+        onlyWhenWalletUnlocked(wallet)
+    {
+        (uint creationFee, uint updateFee, , ) = exchange.getFees();
+        (uint24 accountId, , ) = exchange.getAccount(wallet);
+
+        uint fee = accountId == 0 ? creationFee : updateFee;
+        checkAndReimburse(wallet, fee, feeToken, feeAmount);
+
+        bytes memory callData = abi.encodeWithSelector(
+            exchange.createOrUpdateAccount.selector,
+            pubKeyX,
+            pubKeyY,
+            permission
+        );
+
+        transact(wallet, address(exchange), fee, callData);
+        emit AccountUpdated(address(exchange), wallet, accountId == 0);
+    }
+
+    function depositToDEX(
+        address payable wallet,
+        IExchangeV3     exchange,
+        address         token,
+        uint96          amount,
+        address         feeToken,
+        uint            feeAmount
+        )
+        external
+        nonReentrantExceptFromThis
+        onlyFromMetaTxOrWalletOwner(wallet)
+        onlyWhenWalletUnlocked(wallet)
+    {
+        require(amount > 0, "ZERO_VALUE");
+        (, , uint fee, ) = exchange.getFees();
+        checkAndReimburse(wallet, fee, feeToken, feeAmount);
+
+        bytes memory callData = abi.encodeWithSelector(
+            exchange.deposit.selector,
+            token,
+            amount
+        );
+
+        transact(wallet, address(exchange), fee, callData);
+        emit Deposit(address(exchange), wallet, token, amount);
+    }
+
+    function withdrawFromDEX(
+        address payable wallet,
+        IExchangeV3     exchange,
+        address         token,
+        uint96          amount,
+        address         feeToken,
+        uint            feeAmount
+        )
+        external
+        nonReentrantExceptFromThis
+        onlyFromMetaTxOrWalletOwner(wallet)
+        onlyWhenWalletUnlocked(wallet)
+    {
+        require(amount > 0, "ZERO_VALUE");
+        (, , , uint fee) = exchange.getFees();
+        checkAndReimburse(wallet, fee, feeToken, feeAmount);
+
+        bytes memory callData = abi.encodeWithSelector(
+            exchange.withdraw.selector,
+            token,
+            amount
+        );
+
+        transact(wallet, address(exchange), fee, callData);
+        emit Withdrawal(address(exchange), wallet, token, amount);
+    }
+
+    function checkAndReimburse(
+        address payable wallet,
+        uint            etherFee,
+        address         feeToken,
+        uint            feeAmount
+        )
+        private
+    {
+        if (wallet.balance >= etherFee) {
+            msg.sender.transfer(msg.value);
```

Comment:
`executeMetaTx` will have to be payable so the ETH can be received in the module.

Reply:
I think the ETH will also have to be resent in the `call` so that `msg.value` is the correct value. I think it will be `0` now because the actual `msg.value` is lost in `executeMetaTx` when doing a new `call`.